File: android/app/src/main/java/com/commutescout/drive/Prefs.kt
package com.commutescout.drive

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import java.util.Locale

/**
 * Everything a driver can set. Stored per device; each value is Compose
 * state so screens update the moment it changes.
 */
class Prefs(context: Context) {
    private val p: SharedPreferences = context.getSharedPreferences("cs.prefs", Context.MODE_PRIVATE)

    enum class Theme(val label: String) { SYSTEM("System"), LIGHT("Light"), DARK("Dark") }
    enum class MapStyle(val label: String) {
        AUTO("Match theme"), LIGHT("Light"), DARK("Dark"), OUTDOORS("Outdoors");

        fun serverStyle(dark: Boolean): String = when (this) {
            AUTO -> if (dark) "alidade_smooth_dark" else "alidade_smooth"
            LIGHT -> "alidade_smooth"
            DARK -> "alidade_smooth_dark"
            OUTDOORS -> "outdoors"
        }
    }

    data class LayerKind(val key: String, val label: String, val api: String)

    companion object {
        /**
         * Every layer the website offers. The api name is what
         * /api/mapdata's kinds parameter takes, which is not always the
         * kind the server sends back: closure returns lane_closure,
         * chain returns chain_control and fire returns wildfire.
         */
        val layerKinds = listOf(
            LayerKind("incident", "Incidents", "incident"),
            LayerKind("lane_closure", "Closures and lane work", "closure"),
            LayerKind("chain_control", "Chain controls", "chain"),
            LayerKind("wildfire", "Wildfires", "fire"),
            LayerKind("toll", "Toll prices", "toll"),
            LayerKind("plugin", "Community reports", "plugin"),
            LayerKind("rwis", "Weather stations", "rwis"),
            LayerKind("sign", "Message signs", "sign"),
            LayerKind("camera", "Cameras", "camera"),
        )

        /** The two the website starts with switched off, being large and noisy. */
        val offByDefault = setOf("toll", "camera")

        /**
         * Kinds the app never asks for per viewport. Nationwide they
         * are tens of thousands of points and they carry no line
         * geometry, so the session snapshot is the whole story, as on
         * the website.
         */
        val snapshotOnlyKinds = setOf("camera", "sign", "rwis")
    }

    /** Back to defaults, for UI tests that must start the same way every run. */
    fun resetForTests() {
        p.edit().clear().apply()
        theme = Theme.SYSTEM; mapStyle = MapStyle.AUTO; is3D = true; traffic = false; hiddenKinds = offByDefault
        spokenAlerts = true; alertAheadMeters = 1500.0; showSpeedLimit = true; keepAwake = true
        avoidTolls = false; avoidHighways = false; avoidFerries = false; stripAheadMeters = 16093.0
        advancedAlerts = false; alertRulesRaw = ""; useMiles = true
    }

    var theme by state(Theme.valueOf(p.getString("theme", "SYSTEM")!!)) { p.edit().putString("theme", it.name).apply() }
    var mapStyle by state(MapStyle.valueOf(p.getString("mapstyle", "AUTO")!!)) { p.edit().putString("mapstyle", it.name).apply() }
    var is3D by state(p.getBoolean("3d", true)) { p.edit().putBoolean("3d", it).apply() }
    var traffic by state(p.getBoolean("traffic", false)) { p.edit().putBoolean("traffic", it).apply() }
    var hiddenKinds by state(savedHiddenKinds()) { p.edit().putStringSet("layers.off", it).apply() }
    var spokenAlerts by state(p.getBoolean("spokenalerts", true)) { p.edit().putBoolean("spokenalerts", it).apply() }
    var alertAheadMeters by state(p.getFloat("alertahead", 1500f).toDouble()) { p.edit().putFloat("alertahead", it.toFloat()).apply() }
    var showSpeedLimit by state(p.getBoolean("speedlimit", true)) { p.edit().putBoolean("speedlimit", it).apply() }
    var keepAwake by state(p.getBoolean("keepawake", true)) { p.edit().putBoolean("keepawake", it).apply() }
    var avoidTolls by state(p.getBoolean("avoid.tolls", false)) { p.edit().putBoolean("avoid.tolls", it).apply() }
    var avoidHighways by state(p.getBoolean("avoid.highways", false)) { p.edit().putBoolean("avoid.highways", it).apply() }
    var avoidFerries by state(p.getBoolean("avoid.ferries", false)) { p.edit().putBoolean("avoid.ferries", it).apply() }
    var stripAheadMeters by state(p.getFloat("alerts.strip", 16093f).toDouble()) { p.edit().putFloat("alerts.strip", it.toFloat()).apply() }
    var advancedAlerts by state(p.getBoolean("alerts.advanced", false)) { p.edit().putBoolean("alerts.advanced", it).apply() }
    var alertRulesRaw by state(p.getString("alerts.rules", "")!!) { p.edit().putString("alerts.rules", it).apply() }
    var useMiles by state(if (p.contains("miles")) p.getBoolean("miles", true) else localeMiles()) { p.edit().putBoolean("miles", it).apply() }

    /**
     * Roughly where the driver was when the app last held a snapshot.
     *
     * It is read at launch to aim the next one, so the dots start
     * loading immediately rather than waiting for a location fix. It
     * moves only when the driver leaves the area a snapshot covers,
     * which is hundreds of kilometres wide, so this is not a trail.
     */
    var lastCenter: LatLon?
        get() {
            if (!p.contains("last.lat")) return null
            return LatLon(p.getFloat("last.lat", 0f).toDouble(), p.getFloat("last.lon", 0f).toDouble())
        }
        set(v) {
            if (v == null) p.edit().remove("last.lat").remove("last.lon").apply()
            else p.edit().putFloat("last.lat", v.lat.toFloat()).putFloat("last.lon", v.lon.toFloat()).apply()
        }

    /**
     * Which layers start switched off.
     *
     * A layer added after a release cannot appear in a set saved by an
     * older one, so someone upgrading would find every new layer on,
     * cameras included. The ones the website starts off are added to
     * what is hidden, once, and never again after that.
     */
    private fun savedHiddenKinds(): Set<String> {
        val saved = p.getStringSet("layers.off", null)?.toSet() ?: return offByDefault
        if (p.getBoolean("layers.roadside", false)) return saved
        p.edit().putBoolean("layers.roadside", true).apply()
        return saved + offByDefault
    }

    fun isShown(kind: String) = kind !in hiddenKinds
    fun setShown(kind: String, on: Boolean) { hiddenKinds = if (on) hiddenKinds - kind else hiddenKinds + kind }

    /**
     * How one kind of alert is announced: at what distance, whether it
     * repeats closer, and whether it is spoken at all.
     */
    @kotlinx.serialization.Serializable
    data class AlertRule(val enabled: Boolean = true, val speak: Boolean = true, val firstMeters: Double = 1500.0, val repeatMeters: Double = 0.0)

    val alertKinds = listOf(
        "incident" to "Incidents", "lane_closure" to "Closures and lane work", "chain_control" to "Chain controls",
        "wildfire" to "Wildfires", "police" to "Police reports", "hazard" to "Hazard and crash reports", "plugin" to "Other community reports",
    )

    var alertRules: Map<String, AlertRule>
        get() = runCatching { Backend.json.decodeFromString<Map<String, AlertRule>>(alertRulesRaw) }.getOrDefault(emptyMap())
        set(v) { alertRulesRaw = Backend.json.encodeToString(kotlinx.serialization.serializer<Map<String, AlertRule>>(), v) }

    fun rule(kind: String): AlertRule =
        if (!advancedAlerts) AlertRule(true, spokenAlerts, alertAheadMeters, 0.0) else alertRules[kind] ?: AlertRule()

    fun setRule(kind: String, r: AlertRule) { alertRules = alertRules + (kind to r) }

    /** The rule group for a marker: community reports split by what they are. */
    fun ruleKind(m: RoadMarker): String {
        if (m.kind != "plugin") return m.kind
        val k = (m.flare_kind ?: "").uppercase()
        return when {
            k.startsWith("POLICE") -> "police"
            k.startsWith("HAZARD") || k.startsWith("CRASH") -> "hazard"
            else -> "plugin"
        }
    }

    /**
     * The kinds parameter for /api/mapdata for what is switched on.
     * The snapshot-only kinds stay out of it: asking for them per
     * viewport would triple the size of a call whose real job is to
     * bring back the closure and toll lines the snapshot drops.
     */
    val apiKinds: String get() = layerKinds
        .filter { it.key !in snapshotOnlyKinds && isShown(it.key) }
        .joinToString(",") { it.api }

    /** Valhalla costing options from the route settings; empty when defaults. */
    val costingOptions: Map<String, Any>
        get() {
            val auto = mutableMapOf<String, Any>()
            if (avoidTolls) auto["use_tolls"] = 0
            if (avoidHighways) auto["use_highways"] = 0
            if (avoidFerries) auto["use_ferry"] = 0
            return if (auto.isEmpty()) emptyMap() else mapOf("auto" to auto)
        }

    /** A fingerprint of everything that changes how routes are asked for. */
    val routingKey: String get() = "$avoidTolls|$avoidHighways|$avoidFerries|$useMiles"

    private fun localeMiles(): Boolean = Locale.getDefault().country in setOf("US", "GB", "LR", "MM")

    private fun <T> state(initial: T, save: (T) -> Unit) = object : kotlin.properties.ReadWriteProperty<Any?, T> {
        private var s = mutableStateOf(initial)
        override fun getValue(thisRef: Any?, property: kotlin.reflect.KProperty<*>): T = s.value
        override fun setValue(thisRef: Any?, property: kotlin.reflect.KProperty<*>, value: T) { s.value = value; save(value) }
    }
}

File: ios/CommuteScoutDrive/Prefs.swift
import Combine
import Foundation
import SwiftUI

/// Everything a driver can set. Stored per device in UserDefaults; the
/// keys are stable so nothing is lost across updates.
@MainActor
final class Prefs: ObservableObject {
    enum Theme: String, CaseIterable, Identifiable {
        case system, light, dark
        var id: String { rawValue }
        var label: String { switch self { case .system: "System"; case .light: "Light"; case .dark: "Dark" } }
        var colorScheme: ColorScheme? { switch self { case .system: nil; case .light: .light; case .dark: .dark } }
    }

    enum MapStyle: String, CaseIterable, Identifiable {
        case auto, light, dark, outdoors
        var id: String { rawValue }
        var label: String { switch self { case .auto: "Match theme"; case .light: "Light"; case .dark: "Dark"; case .outdoors: "Outdoors" } }
        /// The server's style name for this choice in the given appearance.
        func serverStyle(dark isDark: Bool) -> String {
            switch self {
            case .auto: isDark ? "alidade_smooth_dark" : "alidade_smooth"
            case .light: "alidade_smooth"
            case .dark: "alidade_smooth_dark"
            case .outdoors: "outdoors"
            }
        }
    }

    /// The marker kinds the map can show, in the order of the Layers sheet.
    static let layerKinds: [(key: String, label: String, api: String)] = [
        ("incident", "Incidents", "incident"),
        ("lane_closure", "Closures and lane work", "closure"),
        ("chain_control", "Chain controls", "chain"),
        ("wildfire", "Wildfires", "fire"),
        ("plugin", "Community reports", "plugin"),
    ]

    @AppStorage("cs.theme") var themeRaw: String = Theme.system.rawValue
    @AppStorage("cs.mapstyle") var mapStyleRaw: String = MapStyle.auto.rawValue
    @AppStorage("cs.3d") var is3D: Bool = true
    @AppStorage("cs.traffic") var traffic: Bool = false
    @AppStorage("cs.layers.off") var layersOff: String = ""      // comma-separated kinds hidden
    @AppStorage("cs.spokenalerts") var spokenAlerts: Bool = true
    @AppStorage("cs.alertahead") var alertAheadMeters: Double = 1500
    @AppStorage("cs.speedlimit") var showSpeedLimit: Bool = true
    @AppStorage("cs.keepawake") var keepAwake: Bool = true
    @AppStorage("cs.avoid.tolls") var avoidTolls: Bool = false
    @AppStorage("cs.avoid.highways") var avoidHighways: Bool = false
    @AppStorage("cs.avoid.ferries") var avoidFerries: Bool = false
    @AppStorage("cs.units") var unitsRaw: String = ""              // "", "mi" or "km"
    @AppStorage("cs.alerts.rules") var alertRulesRaw: String = ""   // JSON, per kind
    @AppStorage("cs.alerts.advanced") var advancedAlerts: Bool = false

    /// How one kind of alert is announced: at what distance, whether it
    /// repeats closer, and whether it is spoken at all. The advanced
    /// mode edits these per kind; the simple mode uses one distance.
    struct AlertRule: Codable, Equatable {
        var enabled = true
        var speak = true
        var firstMeters = 1500.0
        var repeatMeters = 0.0       // 0 = no second warning
    }

    /// The kinds a rule can be set for: the road data kinds plus the
    /// community report groups drivers care about most.
    static let alertKinds: [(key: String, label: String)] = [
        ("incident", "Incidents"),
        ("lane_closure", "Closures and lane work"),
        ("chain_control", "Chain controls"),
        ("wildfire", "Wildfires"),
        ("police", "Police reports"),
        ("hazard", "Hazard and crash reports"),
        ("plugin", "Other community reports"),
    ]

    var alertRules: [String: AlertRule] {
        get {
            guard let d = alertRulesRaw.data(using: .utf8), let r = try? JSONDecoder().decode([String: AlertRule].self, from: d) else { return [:] }
            return r
        }
        set {
            if let d = try? JSONEncoder().encode(newValue), let s = String(data: d, encoding: .utf8) { alertRulesRaw = s }
            objectWillChange.send()
        }
    }

    func rule(for kind: String) -> AlertRule {
        if !advancedAlerts { return AlertRule(enabled: true, speak: spokenAlerts, firstMeters: alertAheadMeters, repeatMeters: 0) }
        return alertRules[kind] ?? AlertRule()
    }

    func setRule(_ r: AlertRule, for kind: String) {
        var all = alertRules
        all[kind] = r
        alertRules = all
    }

    /// The rule group for a marker: community reports split by what they are.
    static func ruleKind(for m: RoadMarker) -> String {
        if m.kind == "plugin" {
            let k = (m.flareKind ?? "").uppercased()
            if k.hasPrefix("POLICE") { return "police" }
            if k.hasPrefix("HAZARD") || k.hasPrefix("CRASH") { return "hazard" }
            return "plugin"
        }
        return m.kind
    }

    var theme: Theme {
        get { Theme(rawValue: themeRaw) ?? .system }
        set { themeRaw = newValue.rawValue; objectWillChange.send() }
    }

    var mapStyle: MapStyle {
        get { MapStyle(rawValue: mapStyleRaw) ?? .auto }
        set { mapStyleRaw = newValue.rawValue; objectWillChange.send() }
    }

    var hiddenKinds: Set<String> {
        get { Set(layersOff.split(separator: ",").map(String.init)) }
        set { layersOff = newValue.sorted().joined(separator: ","); objectWillChange.send() }
    }

    func isShown(_ kind: String) -> Bool { !hiddenKinds.contains(kind) }

    func setShown(_ kind: String, _ on: Bool) {
        var h = hiddenKinds
        if on { h.remove(kind) } else { h.insert(kind) }
        hiddenKinds = h
    }

    /// The kinds parameter for /api/mapdata for what is switched on.
    var apiKinds: String {
        Self.layerKinds.filter { isShown($0.key) }.map(\.api).joined(separator: ",")
    }

    /// Valhalla costing options from the route settings; empty when defaults.
    var costingOptions: [String: Any] {
        var auto: [String: Any] = [:]
        if avoidTolls { auto["use_tolls"] = 0 }
        if avoidHighways { auto["use_highways"] = 0 }
        if avoidFerries { auto["use_ferry"] = 0 }
        return auto.isEmpty ? [:] : ["auto": auto]
    }

    /// A fingerprint of everything that changes how routes are asked for.
    var routingKey: String { "\(avoidTolls)|\(avoidHighways)|\(avoidFerries)|\(Units.useMiles)" }
}
